cmd/mqtt-log-stdout/main.go

```diff
@@ -47,41 +52,73 @@ func start(cfg config.Client) error {
 	metricsServer := newMetricsServer(cfg, statusClient)
 	mqttClient := newMqttClient(cfg, statusClient, messageClient)
 
-	go metricsServer.Start()
+	start(ctx, g, metricsServer)
+	start(ctx, g, mqttClient)
 
-	var result error
-	err := mqttClient.Start()
-	if err != nil {
-		statusClient.Print("Received error starting mqtt client", err)
-		result = multierror.Append(result, err)
-	}
+	stoppedBy := waitForStop(stopChan, ctx)
+	statusClient.Print(fmt.Sprintf("Application stopping, initiated by: %s", stoppedBy), nil)
 
-	stoppedBy := func() string {
-		select {
-		case sig := <-stopChan:
-			return fmt.Sprintf("os.Signal (%s)", sig)
-		case <-mqttClient.Done():
-			return "mqtt client"
-		case <-metricsServer.Done():
-			return "metrics server"
-		}
-	}()
+	cancel()
```

Comment:
Very interesting. In this case, doing cancel here will actually stop the block in the mqttClient - but not stop the actual client. This to make sure that if something goes wrong at startup, it will return an error to the errgroup.

I think it's necessary, as of the current implementation, to call `cancel()` here. But there may be a way around that. I have to think about this.

Reply:
> You deferred `cancel()` above. Are you sure you need an explicit call here?

I have been thinking about this and can't find the correct answer. But my reasoning for keeping the `defer cancel()` in the beginning is to make sure that if something panics it will be cancelled. It "could" also be a protection for the future if there is an early return inserted into the function before the actual `cancel()`.

> Or should this in fact be split into two functions?

No, since I need the context within the errgroup and I want to use the same errgroup between `Start` and `Stop` - or else I won't return an error if an error happens during `Start`. Or at least I think so..

> There seems not to be any early returns so you could just remove the defer, but I think it is good practice to use defer in this case.

That was my thought as well here, doesn't hurt to always use `defer` on these.

Do you agree with my reasoning?